Add document for running JS IPFS in the browser (#2392)
ipfs/js-ipfs: doc/using-js-ipfs-in-the-browser.md

Using JS IPFS in The Browser
----------------------------

JS IPFS is the implementation of IPFS protocol in JavaScript. It can run on any
evergreen browser, inside a service or web worker, as a web extention and in Node.js.

The documentation details more information about running JS IPFS in the browser and
boosting the speed with reducing the errors you face often because of limitation of
js-ipfs.

There are few different components for implementing JS IPFS in the browser. These

> are webrtc and websocket-star which will make it easier for implemented JS IPFS

Author: Having grammar fixes: 873d147a8eae196cfeb67ba1f04a492758bbe534.

in the browser and boost the speed and reduce the errors.

You can find the information present for enabling webRTC support for js-ipfs in
the browser [here](https://github.com/ipfs/js-ipfs#how-to-enable-webrtc-support-for-js-ipfs-in-the-browser) and a different one is
implementing with webrtc-star, you can find information [here](https://github.com/ipfs/js-ipfs#is-there-a-more-stable-alternative-to-webrtc-star-that-offers-a-similar-functionality).

You can find information about running IPFS in the broswer [here](https://github.com/ipfs/js-ipfs#table-of-contents).

Getting Data From IPFS
-----------------------

Using Javascript in the browser

```js
const node = new IPFS()

node.once('ready', () => {
  node.pipe('QmPChd2hVbrJ6bfo3WBcTW4iZnpHm8TEzWkLHmLpXhF68A', (err, data) => {
    if (err) return console.error(err)

    // convert Buffer back to string
    console.log(data.toString())
  })
})
```

JS IPFS Restriction in Web Browser
------------------------------------------

> - DHT in js-ipfs is not by default.

Reviewer: ```suggestion
- DHT in js-ipfs is not enabled by default.
```

> - DHT in js-ipfs is not enabled by default.

Author: ref: https://github.com/ipfs/js-ipfs/pull/2392#discussion_r333407725.

  In browser a JS IPFS node is more restricted because of the same origin policy
  that decrease the discover/connect with peers, external peers without rendezvous
  and relay servers, delegating peer/content routing and preload servers for content

>   discoverablilty hard and erroneous even with swarm peers.

Reviewer 2: Let's rephrase this section to focus on discovery and connectivity, something like:

> This means js-ipfs running in the browser is limited to Web APIs available on a web page.
> There is no raw TCP sockets nor low level UDP, only WebRTC and Websockets.
> Discovery and connectivity to other peers is achieved with a mix of rendezvous and relay servers,
> delegated peer/content routing and preload servers.

- If the browser lacks TCP, you can set up delegating routing. The issue might still
  persist for connecting most of peers in the swarm as described above.

- Connection with websocket ports should be adhering with [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) limitations.

JS IPFS Best Practices
----------------------

- Configure nodes for using webrtc or websocket-star.

> - Run independant instance as default one are under high load and YMMV.

Reviewer 2: ```suggestion
- Run your own instance of `*-star` signaling service. The default ones are under high load and YMMV.
```

- Make sure content you are implementing is cached in go-ipfs
  - manually `pin` or preload CIDs of interest with `refs -r` beforehand.
  - preload content on the fly using [preload](https://github.com/ipfs/js-ipfs#optionspreload) or

>     configuring [delgated routing](https://github.com/ipfs/js-ipfs#configuring-delegate-routers).

Reviewer: ```suggestion
    configuring [delegated routing](https://github.com/ipfs/js-ipfs#configuring-delegate-routers).
```

js-ipfs.

There are few different components for implementing JS IPFS in the browser. These
are webrtc and websocket-star which will make it easier for implemented JS IPFS
in the browser and boost the speed and reduce the errors.

You can find the information present for enabling webRTC support for js-ipfs in
the browser [here](https://github.com/ipfs/js-ipfs#how-to-enable-webrtc-support-for-js-ipfs-in-the-browser) and a different one is
implementing with webrtc-star, you can find information [here](https://github.com/ipfs/js-ipfs#is-there-a-more-stable-alternative-to-webrtc-star-that-offers-a-similar-functionality).

You can find information about running IPFS in the broswer [here](https://github.com/ipfs/js-ipfs#table-of-contents).

Getting Data From IPFS
-----------------------

Using Javascript in the browser

```js
const node = new IPFS()

node.once('ready', () => {
  node.pipe('QmPChd2hVbrJ6bfo3WBcTW4iZnpHm8TEzWkLHmLpXhF68A', (err, data) => {
    if (err) return console.error(err)

    // convert Buffer back to string
    console.log(data.toString())
  })
})
```

JS IPFS Restriction in Web Browser
------------------------------------------

> - There is no DHT in js-ipfs. 

Reviewer: This is no longer accurate. There is a DHT just not enabled by default. This PR is working on integrating it by default https://github.com/ipfs/js-ipfs/pull/1994

  In browser a JS IPFS node is more restricted because of the same origin policy
  that decrease the discover/connect with peers, external peers without rendezvous
  and relay servers, delegating peer/content routing and preload servers for content
  discoverablilty hard and erroneous even with swarm peers.

- Unable in opening TCP connections even if one sets up delegating routing the 
issue will persist for connecting most of peers in the swarm as described a
little above.

- Connection with websocket ports should be adhering with [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) limitations.

JS IPFS Best Practices
----------------------

> - Run in fedrated model.

Reviewer: @tapaswenipathak what do you mean by this?

Author: I don't remember now.
